src/main/java/io/anserini/rerank/lib/RocchioReranker.java

```diff
@@ -0,0 +1,212 @@
+  }
+
+  @Override
+  public ScoredDocuments rerank(ScoredDocuments docs, RerankerContext context) {
+    assert(docs.documents.length == docs.scores.length);
+
+    IndexSearcher searcher = context.getIndexSearcher();
+    IndexReader reader = searcher.getIndexReader();
+
+    FeatureVector qfv = FeatureVector.fromTerms(AnalyzerUtils.analyze(analyzer, context.getQueryText()));
+
+    FeatureVector rm = computeMeanOfDocumentVectors(docs, reader);
+
+    // Rocchio weight = alpha * original query + beta * mean(top n doc embedding)
+    // Here beta = 1-alpha
+    rm = interpolateQueryDocVector(qfv, rm, alpha, beta);
+
+    BooleanQuery.Builder feedbackQueryBuilder = new BooleanQuery.Builder();
+
+    Iterator<String> terms = rm.iterator();
+    while (terms.hasNext()) {
+      String term = terms.next();
+      float prob = rm.getFeatureWeight(term);
+      feedbackQueryBuilder.add(new BoostQuery(new TermQuery(new Term(this.field, term)), prob), BooleanClause.Occur.SHOULD);
+    }
+
+    Query feedbackQuery = feedbackQueryBuilder.build();
+
+    if (this.outputQuery) {
+      LOG.info("QID: " + context.getQueryId());
+      LOG.info("Original Query: " + context.getQuery().toString(this.field));
+      LOG.info("Running new query: " + feedbackQuery.toString(this.field));
+    }
+
+    TopDocs rs;
+    try {
+      Query finalQuery = feedbackQuery;
+      // If there's a filter condition, we need to add in the constraint.
+      // Otherwise, just use the feedback query.
+      if (context.getFilter() != null) {
+        BooleanQuery.Builder bqBuilder = new BooleanQuery.Builder();
+        bqBuilder.add(context.getFilter(), BooleanClause.Occur.FILTER);
+        bqBuilder.add(feedbackQuery, BooleanClause.Occur.MUST);
+        finalQuery = bqBuilder.build();
+      }
+
+      // Figure out how to break the scoring ties.
+      if (context.getSearchArgs().arbitraryScoreTieBreak) {
+        rs = searcher.search(finalQuery, context.getSearchArgs().hits);
+      } else if (context.getSearchArgs().searchtweets) {
+        rs = searcher.search(finalQuery, context.getSearchArgs().hits, BREAK_SCORE_TIES_BY_TWEETID, true);
+      } else {
+        rs = searcher.search(finalQuery, context.getSearchArgs().hits, BREAK_SCORE_TIES_BY_DOCID, true);
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
+      return docs;
+    }
+
+    return ScoredDocuments.fromTopDocs(rs, searcher);
+  }
+
+  private FeatureVector computeMeanOfDocumentVectors(ScoredDocuments docs, IndexReader reader) {
+    FeatureVector f = new FeatureVector();
+
+    Set<String> vocab = new HashSet<>();
+    int numdocs;
+    numdocs = docs.documents.length < fbDocs ? docs.documents.length : fbDocs;
+
+    List<FeatureVector> docvectors = new ArrayList<>();
+    for (int i = 0; i < numdocs; i++) {
+      try {
+        FeatureVector docVector = createDocumentVector(reader.getTermVector(docs.ids[i], field), reader, docs.ids[i]);
+        docVector.pruneToSize(fbTerms);
```

Comment:
I don't think we should prune each doc vector to size... since we already prune at the end.